OGM-1038 Pass transaction id to dialects

DavideD opened the pull request:
https://hibernate.atlassian.net/browse/OGM-1038

This will make it possible for a dialect to access information about the current transaction. At the moment is only the transaction id.

I haven't added any tests for this specific part since none of the current dialects need this information.
It will be used by the Neo4j remote dialect in following PR.

Note that I added a `QueryContext` to the method executing native queries. It will also store the `TupleContext` removing the need to save the sessionFactory in the Neo4j dialect.

We already discussed about this PR in the pasts but we changed the codebase a bit so, maybe, you won't to have a quick look.

gsmet commented on core/src/main/java/org/hibernate/ogm/dialect/spi/TransactionContext.java:
```diff
@@ -0,0 +1,15 @@
+package org.hibernate.ogm.dialect.spi;
+
+/**
+ * @author Davide D'Alto
```
Maybe a small javadoc would be nice?

gsmet commented on core/src/main/java/org/hibernate/ogm/dialect/impl/EmptyTransactionContext.java:
```diff
@@ -0,0 +1,32 @@
+
+	@Override
+	public Object getTransactionId() {
+		throw LOG.transactionIdIsNull();
```
I would have named it `transactionIdIsNotAvailable` as the message of the exception states it.

gsmet commented:
Looks good to me overall apart from the minor nitpicking.

That being said, not sure I understand all the implications of this so I would prefer to have a second opinion before merging.

DavideD commented:
I've applied @gsmet remarks

gunnarmorling commented on core/src/main/java/org/hibernate/ogm/persister/impl/OgmEntityPersister.java:
```diff
@@ -1743,8 +1746,17 @@ public String getSubclassForDiscriminatorValue(Object value) {
+	 * @param session the current session, if null the {@link TupleContext#getTransactionContext()} will be null.
+	 * @return the tupleContext for the session
+	 */
+	public TupleContext getTupleContext( SessionImplementor session ) {
```
When will the session be null?

DavideD replied:
Good question ... checking ...

DavideD replied:
It happens when massindexing, in that case there is not a session when the foreach method is called.
At the moment, Neo4j does not work with the massindexer because of different Lucene versions. 
It should be possible to implement it in the remote scenario but I haven't done that yet

gunnarmorling commented on core/src/main/java/org/hibernate/ogm/util/impl/TupleContextHelper.java:
```diff
@@ -0,0 +1,34 @@
+	 * EntityMetadataInformation is {@code null}
+	 */
+	public static TupleContext tupleContext(SessionImplementor session, EntityMetadataInformation metadata) {
+		if ( metadata != null ) {
```
When would that be the case? Seems like null should be prevented in the first place?

DavideD replied:
Check `BackendQuery#getSingleEntityMetadataInformationOrNull()`:
```
	/**
	 * Returns the {@link EntityMetadataInformation} of the entity type selected by this query.
	 *
	 * @return the {@link EntityMetadataInformation} of the entity type selected by this query or {@code null} in case this
	 * query does not select exactly one entity type (e.g. in case of scalar values or joins (if supported in future revisions)).
	 */
	public EntityMetadataInformation getSingleEntityMetadataInformationOrNull() {
		return singleEntityMetadataInformation;
	}

```

gsmet commented:
As discussed on Thursday, you need to check if QueryContext is required as the TransactionContext in it seems redundant with the TransactionContext located in the TupleContext.

DavideD commented:
I removed the `QueryContext` interface and rebased.

Having a TupleContext in executeBackendQuery also allowed me to remove the SessionFactory from the Neo4j dialect. See [OGM-963](https://hibernate.atlassian.net/browse/OGM-963)

gsmet commented:
Merged! Thanks!
